soc: nordic: common: dmm: Optimize DMM operations

nordic-krch opened the pull request:
DMM is handling bounce buffers used for DMA in peripherals. There are 2 heaps that are used: RAM3 (slow RAM) and RAM2 (fast, cached). DMM operations were extremely slow adding significant overhead to the operation of the drivers. There are 2 reasons for that:
- sys_heap is used for allocation - sys_heap is using control data next to the heap data and for RAM3 where access is slow it results in extremely long heap operations (alloc+free almost 30us)
- standard memcpy is used which by default is slow (see #95154)

PR introduces following optimizations:
- Use optimized memcpy. If `src`, `dst` and `len` are 64 bit aligned do the double-word access, if all are 32 bit aligned do the word access. Fallback to the default memcpy.
- Use local micro heap which divides available heap space into multiply of 32 chunks (by default 32 chunks) and uses 32 bit mask to find a contiguous memory space (chunk granularity).  Default use of 32 chunks should be enough in most of the cases but it is possible to increase that using `CONFIG_ DMM_HEAP_CHUNKS` Kconfig option.
- DMM test suite measures time spend in each operation
- Added an option optional function `dmm_stats_get` to track DMM usage statistics (`CONFIG_DMM_STATS=y`).

Timing improvements for aligned buffers where optimized heap and memcpy is used. 

RAM3 approx. 10x improvement for 16B and 5x for 128B:

| Operation | before (cyc) | after (cyc) |
| ------------- | --------- | ------ |
| out_prepare 16B | 6598 | 873 |
| out_release 16B | 4966 | 255 |
| out_prepare 128B | 20030 | 4234 |
| out_release 128B | 4986 | 255 |
| in_prepare 16B | 4669 | 386 |
| in_release 16B | 6607 | 699 |
| in_prepare 128B | 4723 | 363 |
| in_release 128B | 17769 | 3513 |

RAM2 - approx. 3x improvemen: 

| Operation | before (cyc) | after (cyc) |
| ------------- | --------- | ------ |
| out_prepare 16B | 1254 | 510 |
| out_release 16B | 1091 | 252 |
| out_prepare 128B | 2309 | 914 |
| out_release 128B | 1090 | 252 |
| in_prepare 16B | 1110 | 465 |
| in_release 16B | 1349 | 453 |
| in_prepare 128B | 1313 | 517 |
| in_release 128B | 2763 | 1103 |

~~DNM until #95416 is merged.~~

nordic-krch committed:
tests: boards: nrf: dmm: Add timing measurements

Add timing measurement to the test to allow DMM profiling.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: boards: nrf: dmm: Align test to changes in DMM

Align to API changes.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
soc: nordic: common: dmm: Optimize memcpy

Default memcpy used in zephyr is not optimized and performs simple
byte by byte copying. Using double word or word access can significantly
reduce copying time especially for RAM3 (slow peripheral RAM).

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
soc: nordic: common: dmm: Optimize by using a micro heap

Add micro heap implementation which is using one or more 32 bit masks
to allocate quickly blocks. It is significantly better than using
sys_heap. Difference is especially big on RAM3 heap because heap
control data is in RAM3 space so operations there were extremely
slowly (15 us to allocate a buffer).

Simplified implementation of the heap requires DMM API change as
release functions need to know the length of the allocated buffer as
simple heap requires that (buffer address is enough for the standard
heap).

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

bjarki-andreasen reviewed (changes requested):
The DMM optimization is great, but the addition to the sys_bitarray API should be refactored to work with bitarrays, and that should be used in dmm.c

nashif commented on include/zephyr/sys/bitarray.h:
```diff
@@ -146,6 +146,23 @@ int sys_bitarray_test_and_set_bit(sys_bitarray_t *bitarray, size_t bit, int *pre
  */
 int sys_bitarray_test_and_clear_bit(sys_bitarray_t *bitarray, size_t bit, int *prev_val);
 
+/**
```
such changes to common code and APIs should be submitted standalone and should not be mixed with SoC and platform changes, this makes it go fast through the review. Mixing things like that just make this almost impossible to move forrward, i.e. I have now to approve both common code and SoC code I do not understand and I am not qualified to review.

nordic-krch replied:
Bitarray changes are now in #95416

nordic-krch commented:
Timings updated as they slightly worsened after moving to use bitarray in DMM (+50-80 cycles per operation).

nordic-krch commented:
@bjarki-andreasen can you take another look as bitarray part is now merged?
@nika-nordic can you take a look?

nika-nordic commented on soc/nordic/common/dmm.h:
```diff
@@ -111,11 +111,12 @@ int dmm_buffer_out_prepare(void *region, void const *user_buffer, size_t user_le
  * @param region Memory region associated with device to release the buffer for.
  * @param buffer_out Bus address of the DMA output buffer previously prepared
  *                   with @ref dmm_buffer_out_prepare().
+ * @param user_length Length of the allocated buffer.
```
how the `dmm` user is supposed to know how much data was allocated underneath? `user_length` and `alloc_length` are not equal is `user_length` is smaller than multiple of allocation block size, correct?

Can't be `alloc_length` derived from `user_length` by taking `user_length` and finding next multiple of allocation block size? If yes, then maybe to streamline the API, we could refactor it in the following way:
```
int dmm_buffer_out_prepare(void *region, void const *user_buffer, size_t user_length, void **buffer_out);

int dmm_buffer_out_release(void *region, void *buffer_out, size_t user_length);
// alternatively: int dmm_buffer_out_release(void *region, void const *user_buffer/*UNUSED*/, size_t user_length, void *buffer_out);

int dmm_buffer_in_prepare(void *region, void *user_buffer, size_t user_length, void **buffer_in);

int dmm_buffer_in_release(void *region, void *user_buffer, size_t user_length, void *buffer_in);
```
this way we could get rid of the extra parameter and making the APIs more similar

nika-nordic reviewed (commented):
I have some concerns regarding dmm API usability after `alloc_length` parameter addition. Otherwise LGTM

nordic-krch committed:
soc: nordic: common: dmm: Add optional usage stats

Add support for getting usage statistics for DMM.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: boards: nrf: dmm: Extend test coverage

Extended test to check usage stats and longer buffer alloc.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

bjarki-andreasen reviewed (dismissed):
Looks good

nordic-krch commented:
@anangl can you take another look?

nordic-krch committed:
tests: boards: nrf: dmm: Add stress test

Add stress test which validates that allocator is thread safe
and has no memory leaks.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on soc/nordic/common/dmm.c:
```diff
@@ -103,37 +109,142 @@ static bool is_user_buffer_correctly_preallocated(void const *user_buffer, size_
+	atomic_t *t_mask = &tail_mask[idx];
+
+	mask = (*t_mask | BIT(w_off)) >> w_off;
+	num_bits = __builtin_ctz(~mask);
```
fixed

nordic-krch commented on soc/nordic/common/dmm.c:
```diff
@@ -103,37 +109,142 @@ static bool is_user_buffer_correctly_preallocated(void const *user_buffer, size_
+			num_bits += 32;
+			*t_mask = 0;
+			t_mask++;
+		}
```
reworked it into a common while loop.
